From e2cdb5c4c7733bc45a5536b5a6c2a951c693dbc0 Mon Sep 17 00:00:00 2001
From: Michael Chang <mchang@suse.com>
Date: Mon, 14 Jul 2025 22:10:18 +0800
Subject: [PATCH 2/4] http: Return HTTP status code in http_establish

Previously, using "test -s ..." or "test -f ..." on files served via
HTTP would always return true, regardless of whether the target file
actually existed. This is incorrect behavior, whereas the same tests
work as expected with TFTP.

The issue stems from http_establish returning success (GRUB_ERR_NONE) as
long as the HTTP connection was established, without considering the
HTTP status code returned in the response. As a result, http_open would
always report success, discarding error responses such as 404 Not Found.

The patch makes http_establish to return the HTTP status code as its
return value when an error or unknown status code is encountered. It
also sets data->first_line_recv = 1 in the parse_line()'s error code
path to correctly reflect the parsing state and prevent reprocessing.

With this change, both -s and -f tests now behave correctly when used
with the HTTP protocol, as http_establish returns error for failed HTTP
status codes as well. As a result, http_open is no longer considered
successful solely based on establishing the connection, it now also
takes the HTTP status code into account.

Signed-off-by: Michael Chang <mchang@suse.com>
Reviewed-by: Andrew Hamilton <adhamilt@gmail.com>
Reviewed-by: Avnish Chouhan <avnish@linux.ibm.com>
Reviewed-by: Leo Sandoval <lsandova@redhat.com>
Part-of: <https://gitlab.freedesktop.org/gnu-grub/grub/-/merge_requests/167>
---
 grub-core/net/http.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/grub-core/net/http.c b/grub-core/net/http.c
index 5a19f8d95..94289259c 100644
--- a/grub-core/net/http.c
+++ b/grub-core/net/http.c
@@ -122,6 +122,7 @@ parse_line (grub_file_t file, http_data_t data, char *ptr, grub_size_t len)
 	case 404:
 	  data->err = GRUB_ERR_FILE_NOT_FOUND;
 	  data->errmsg = grub_xasprintf (_("file `%s' not found"), data->filename);
+	  data->first_line_recv = 1;
 	  return GRUB_ERR_NONE;
 	default:
 	  data->err = GRUB_ERR_NET_UNKNOWN_ERROR;
@@ -129,6 +130,7 @@ parse_line (grub_file_t file, http_data_t data, char *ptr, grub_size_t len)
 	     valid answers like 403 will trigger this very generic message.  */
 	  data->errmsg = grub_xasprintf (_("unsupported HTTP error %d: %s"),
 					 code, ptr);
+	  data->first_line_recv = 1;
 	  return GRUB_ERR_NONE;
 	}
       data->first_line_recv = 1;
@@ -473,6 +475,18 @@ http_establish (struct grub_file *file, grub_off_t offset, int initial)
 	}
       return grub_error (GRUB_ERR_TIMEOUT, N_("time out opening `%s'"), data->filename);
     }
+
+  if (data->err)
+    {
+      char *str = data->errmsg;
+      if (data->sock)
+	grub_net_tcp_close (data->sock, GRUB_NET_TCP_ABORT);
+      err = grub_error (data->err, "%s", str);
+      grub_free (str);
+      data->errmsg = 0;
+      return err;
+    }
+
   return GRUB_ERR_NONE;
 }
 
-- 
2.55.0

